Update archive to use github.com/moby/go-archive

dmcgowan opened the pull request:
Updates use of archive of associated uses of idtools.

 Still a few remaining packages using idtools and requires an updated buildkit to fully get rid of idtools.

```markdown changelog
Go-SDK: deprecate `pkg/archive`, which was migrated to `github.com/moby/go-archive`.
```

dmcgowan committed:
deprecate pkg/(chroot)archive for  github.com/moby/go-archive

- pkg/archive: deprecate, and add aliases
  Keeping the tests in this commit; also moves various utilities
  into a _test.go file, as they were now only used in tests.
- pkg/chrootarchive: deprecate and add aliase
  deprecate pkg/archive and add aliases
  keeping the tests in this commit
- Add temporary exceptions for deprecation linting errors, because
  this commit is to verify everything works with the aliases.
- remove tests that depend on un-exported types

    === RUN   TestDisablePigz
    --- FAIL: TestDisablePigz (0.00s)
    panic: interface conversion: io.Reader is *archive.readCloserWrapper, not *archive.readCloserWrapper (types from different packages) [recovered]

- pkg/archive, pkg/chrootarchive: remove test files

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>
Signed-off-by: Derek McGowan <derek@mcg.dev>

dmcgowan committed:
Update to use github.com/moby/go-archive

Update use of idtools to moby/user for archive and other deprecated uses

Signed-off-by: Derek McGowan <derek@mcg.dev>

crazy-max commented:
Can you remove the `toolchain` directive in `go.mod` from upstream: https://github.com/moby/go-archive/blob/13fe27747a57a6b35c912fe21e63f3e087169a2d/go.mod#L5?

dmcgowan commented:
@crazy-max updated

I'm going to mark this as ready for review, there are still a few changes that will happen in follow ups. The archive package needs some CI setup and could possibly do a few more changes such as the compression split before the first tag. We will also need to update this further after the buildkit update, I put TODO in the places we will likely need to update along with buildkit for idtools->user move. @jsternberg maybe pull in #49747 or we can figure out which one goes first.

thaJeztah commented:
We should probably (at least) squash the last two commits, because the second commit doesn't compile.

I also worked on changes to deprecate both `pkg/archive` and `pkg/chrootarchive` and to make them aliases for the new module;

- this variant adds the aliases, but keeps the tests to verify the aliases work without changes elsewhere; https://github.com/moby/moby/pull/49752
- this variant does the same, but discards the tests; https://github.com/moby/moby/pull/49758
- :point_up: :warning: we should make CI work in the https://github.com/moby/go-archive repository, because currently the module is not covered by tests
- finally, I have this PR https://github.com/moby/moby/pull/49759 that squashes the commits, and replaces some remaining uses of ` idtools.MkdirAllAndChown`, and `idtools.MkdirAllAndChownNew`

Combining the "deprecated and alias" with the `vendor` commit makes sure that the migration shows up as a lateral move (with some minor changes), which makes it much more transparent what's done;

<img width="383" alt="Screenshot 2025-04-07 at 10 58 24" src="https://github.com/user-attachments/assets/1d1cb94a-bd82-4278-8648-5f93a166a88c" />

jsternberg commented:
I've created https://github.com/moby/buildkit/pull/5891 to remove this dependency from buildkit and then I'll update #49747.

dmcgowan commented:
> Combining the "deprecated and alias" with the vendor commit makes sure that the migration shows up as a lateral move (with some minor changes), which makes it much more transparent what's done;

Agreed, I don't think we need to include that in this PR though.

Rebased and squashed last two commits.

dmcgowan commented:
>  I'd really rather have it done in a single atomic commit; is there a specific reason you'd prefer not doing it?

You can pull in that commit if you already have it and just want Github to show it as a move.

dmcgowan commented:
@thaJeztah rebased on your vendor/deprecation commit

thaJeztah reviewed (approved):
LGTM, thanks!
